tardis/energy_input/gamma_ray_packet_source.py

```diff
@@ -0,0 +1,298 @@
+import numpy as np
+from numba import njit
+
+from tardis.energy_input.energy_source import (
+    positronium_continuum,
+)
+from tardis.energy_input.GXPacket import (
+    GXPacketCollection,
+    initialize_packet_properties,
+)
+from tardis.energy_input.util import (
+    get_index,
+    get_random_unit_vector,
+    doppler_factor_3d,
+    H_CGS_KEV,
+)
+from tardis.energy_input.samplers import sample_energy
+from tardis.montecarlo.montecarlo_numba import njit_dict_no_parallel
+from tardis.montecarlo.packet_source import BasePacketSource
+
+
+class RadioactivePacketSource(BasePacketSource):
+    def __init__(
+        self,
+        packet_energy,
+        gamma_ray_lines,
+        positronium_fraction,
+        inner_velocities,
+        outer_velocities,
+        inv_volume_time,
+        times,
+        energy_df_rows,
+        effective_times,
+        taus,
+        parents,
+        average_positron_energies,
+        inventories,
+        average_power_per_mass,
+        **kwargs,
+    ):
+        self.packet_energy = packet_energy
+        self.gamma_ray_lines = gamma_ray_lines
+        self.positronium_fraction = positronium_fraction
+        self.inner_velocities = inner_velocities
+        self.outer_velocities = outer_velocities
+        self.inv_volume_time = inv_volume_time
+        self.times = times
+        self.energy_df_rows = energy_df_rows
+        self.effective_times = effective_times
+        self.taus = taus
+        self.parents = parents
+        self.average_positron_energies = average_positron_energies
+        self.inventories = inventories
+        self.average_power_per_mass = average_power_per_mass
+        super().__init__(**kwargs)
+
+    @njit(**njit_dict_no_parallel)
+    def create_packet_radii(
+        self, no_of_packets, inner_velocity, outer_velocity
+    ):
+        """Initialize the random radii of packets in a shell
+
+        Parameters
+        ----------
+        packet_count : int
+            Number of packets in the shell
+        inner_velocity : float
+            Inner velocity of the shell
+        outer_velocity : float
+            Outer velocity of the shell
+
+        Returns
+        -------
+        array
+            Array of length packet_count of random locations in the shell
+        """
+        z = np.random.random(no_of_packets)
+        initial_radii = (
+            z * inner_velocity**3.0 + (1.0 - z) * outer_velocity**3.0
+        ) ** (1.0 / 3.0)
+
+        return initial_radii
+
+    def create_packet_nus(self, no_of_packets, energy, intensity):
+        nu_energies = np.zeros(no_of_packets)
+        for i in range(no_of_packets):
+            nu_energies[i] = sample_energy(energy, intensity)
+        return nu_energies
```

Comment:
Does the sample_energy function gives (energy * times). Is the nu_energies confusing?

Reply:
I agree that it is confusing. One option is to only track nu_rf and nu_cmf on the packet, and just calculate energy_rf and energy_cmf (either with a method on the packet or another way). However we *sample* from energy lists, not frequency lists.

Reply:
Never mind, the packet does only include frequency and its total energy